module/zfs/zfs_log.c

```diff
@@ -499,6 +500,94 @@ zfs_log_rename(zilog_t *zilog, dmu_tx_t *tx, uint64_t txtype,
+ *     RENAME_WHITEOUT: creates a whiteout inode in place of renamed file as
+ *                      an atomic operation
+ *
+ * Ideally, these operations should be represented as new ZFS Intent Log record
+ * types, which should mandate a new ZFS feature flag due to the on-disk format
+ * change. One would use spa_feature_incr/decr functions to indicate that we're
+ * actually actively using the new on-disk txtypes - but these functions are
+ * only supposed to be called from the txg syncing context.
+ * This means that would need to force out in-progress txg to disk and start
+ * a new one before writing any ZIL records, just so we can be sure that ZIL
+ * replaying ZFS gets told it should expect potentially incompatible ZIL
+ * txtypes. Doing this would hurt performance.
+ * Alternatively, we could just activate the feature on a pool when these
+ * renameat2 flags get first used and leave it at that - which would render the
+ * pool importable read-only on implementations without the new feature flag,
+ * even when no new txtypes IL records would be present on-disk - which on most
+ * setups could be 'almost' all the time, so it'd be a shame to have them all
+ * read-only on non-Linux platforms.
+ * As a third option, at least until more platforms implement renameat2, we
+ * choose to rely on the fact that the ZIL is replayed in single-threaded mode
+ * before the dataset is mounted. This way, we can represent the otherwise
+ * atomic operations as a series of plain good old txtypes known to all current
+ * OpenZFS implementations. To do that, we use these hacky functions:
+ *
+ * zfs_log_rename_exchange
+ * zfs_log_rename_whiteout
+ *
+ *     To represent atomic rename with old non-atomic operations, we need
+ *     a temporary new name; so we try picking a name until we succeed, then
+ *     we get a dirent lock for that temp name until the final itx gets queued
+ */
+
+void
+zfs_log_rename_exchange(zilog_t *zilog, dmu_tx_t *tx, uint64_t txtype,
+    znode_t *sdzp, char *sname, znode_t *tdzp, char *dname, znode_t *szp)
+{
+	zfs_dirlock_t *tmpdl;
+	znode_t *tmpzp = NULL;
+	char rndname[16];
+	char *tmpname;
+	int retries = 0;
+	int error;
+
+	tmpname = kmem_alloc(MAXPATHLEN, KM_SLEEP);
+	ASSERT3P(tmpname, !=, NULL);
+
+	for (int i = 0; i < 16; i++) {
+		int r = 0xFF;
+		while (r > 127 || r == 0 || r == '/')
+			random_get_pseudo_bytes((void *)&r, 1);
+		rndname[i] = (char)r;
+	}
+
+	do {
+		retries++;
+		(void) snprintf(tmpname, MAXPATHLEN,
+		    "%s.zfs_renameat2_emul_%s%4d",
+		    dname, rndname, retries);
+		error = zfs_dirent_lock(&tmpdl, tdzp, tmpname,
+		    &tmpzp, ZNEW, NULL, NULL);
+	} while (error != 0 && retries != INT_MAX);
+	ASSERT3U(retries, !=, INT_MAX);
+
+	zfs_log_rename(zilog, tx, txtype, sdzp, sname, tdzp, tmpname, szp);
+	zfs_log_rename(zilog, tx, txtype, tdzp, dname, sdzp, sname, szp);
+	zfs_log_rename(zilog, tx, txtype, tdzp, tmpname, tdzp, dname, szp);
```

Comment:
Nit: I think the ordering should be the more traditional:

```suggestion
	/* dst -> tmp */
	zfs_log_rename(zilog, tx, txtype, tdzp, dname, tdzp, tmpname, szp);
	/* src -> dst */
	zfs_log_rename(zilog, tx, txtype, sdzp, sname, tdzp, dname, szp);
	/* tmp -> src */
	zfs_log_rename(zilog, tx, txtype, tdzp, tmpname, sdzp, sname, szp);
```